src/Derivative.h

```diff
@@ -19,71 +19,37 @@ namespace Halide {
  *  Helper structure storing the adjoints Func.
  *  Use d(func) or d(buffer) to obtain the derivative Func.
  */
-struct Derivative {
+class Derivative {
+public:
     // function name & update_id, for initialization update_id == -1
     using FuncKey = std::pair<std::string, int>;
 
-    std::map<FuncKey, Func> adjoints;
+    explicit Derivative(const std::map<FuncKey, Func> &adjoints_in) : adjoints(adjoints_in) {}
+    explicit Derivative(std::map<FuncKey, Func> &&adjoints_in) : adjoints(std::move(adjoints_in)) {}
 
-    Func get(const Func &func, int update_id = -1, bool bounded = true) const {
-        std::string name = func.name();
-        if (!bounded) {
-            name += "_unbounded";
-        }
-        auto it = adjoints.find(FuncKey{ name, update_id });
-        if (!bounded && it == adjoints.end()) {
-            // No boundary condition applied, use the original function
-            name = func.name();
-            it = adjoints.find(FuncKey{ name, update_id });
-        }
-        internal_assert(it != adjoints.end()) << "Could not find Func " << name << "\n";
-        return it->second;
-    }
+    Func get(const Func &func, int update_id = -1, bool bounded = true) const;
+    Func get(const Buffer<> &buffer) const;
+    Func get(const Param<> &param) const;
 
-    Func operator()(const Func &func, int update_id = -1, bool bounded = true) const {
+    inline Func operator()(const Func &func, int update_id = -1, bool bounded = true) const {
```

Comment:
I thought member functions defined within the class were automatically inline.

Reply:
Open https://github.com/halide/Halide/issues/4264 to track adding tests for the unbounded case.

re: member functions automatically being inlined, yes, this is redundant, I'll eliminate them.

(I'd still like to know why we have both `get()` and `operator()` and whether we can eliminate one or the other.)